ostd/src/mm/dma/dma_coherent.rs

```diff
@@ -160,24 +165,26 @@ impl Drop for DmaCoherentInner {
 }
 
 impl VmIo for DmaCoherent {
-    fn read_bytes(&self, offset: usize, buf: &mut [u8]) -> crate::prelude::Result<()> {
-        self.inner.vm_segment.read_bytes(offset, buf)
+    fn read_bytes(&self, _offset: usize, _buf: &mut [u8]) -> Result<()> {
+        panic!(concat!(
+            "`DmaCoherent` must be accessed with non-tearing memory operations, ",
+            "please consider using `read_val` instead"
```

Comment:
Yes, we should fix `read_val`, but why forbid `read_bytes`? I don't think there is an architectural requirement that forbids all cache-coherent DMA areas from being read/written in a stream fashion. I think it is Virtio that demands accessing the fields of its configuration objects in single memory load/stores, not DMA.

I think we should add a `read_once` method to `SafePtr`.

```rust
impl<T: Pod, M: VmIo, R: TRights> SafePtr<T, M, TRightSet<R>> {
    #[require(R > Read)]
    pub fn read_once(&self) -> Result<T> {
        self.vm_obj.read_once(self.offset)
    }
}
```

And refactor the virtio driver to use the `read_once` method of `SafePtr` instead.

Reply:
> Yes, we should fix read_val as a quick fix, but why forbid read_bytes? I don't think there is an architectural requirement that forbids all cache-coherent DMA areas from being read/written in a stream fashion. I think it is Virtio that demands accessing the fields of its configuration objects in single memory load/stores, not DMA.

I agree with that. But having `read_bytes` and `read_val` have different semantics is weird. As far as I am concerned, the following two should always be the same:
```rust
   reader.read_bytes(64, &mut buffer);
   let bytes: [u8; 64] = reader.read_val();
```

So since there are currently no users of `read_bytes`, changing it to just panic is acceptable to me.

The panic also serves as a note that `DmaCoherent` currently does not support reading/writing byte slices, neither with `read_bytes`/`write_bytes` nor with `read_val`/`write_val`. So future users should fix this problem first.

> A probably-more long-term solution, I think, is to add a `read_once` method to `SafePtr`.
> 
> ```rust
> impl<T: Pod, M: VmIo, R: TRights> SafePtr<T, M, TRightSet<R>> {
>     #[require(R > Read)]
>     pub fn read_once(&self) -> Result<T> {
>         self.vm_obj.read_once(self.offset)
>     }
> }
> ```
> 
> And refactor the virtio driver to use the `read_once` method of `SafePtr` instead.

Yes, I'm thinking about something like that. I can also try the solution in this PR if you like it.

Reply:
> having read_bytes and read_val have different semantics is weird.

I meant that they do have (and should, as you proposed) the same semantics if we have `read_once`, which is different from `read_bytes` and `read_val`. So we can still keep the current implementation of `read_val` and `read_bytes` intact and just add `read_once` and refactor the virtio driver.